cps/kni.c

```diff
@@ -1261,6 +1261,114 @@ rd_modroute(const struct nlmsghdr *req, struct cps_config *cps_conf, int *err)
 	return MNL_CB_OK;
 }
 
+static void
+rd_fill_getaddr_reply(const struct cps_config *cps_conf,
+	struct mnl_nlmsg_batch *batch, struct gatekeeper_if *iface,
+	uint8_t family, uint8_t kni_index, uint32_t seq)
+{
+	struct nlmsghdr *reply;
+	struct ifaddrmsg *ifam;
+
+	reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+	reply->nlmsg_type = RTM_NEWADDR;
+	reply->nlmsg_flags = NLM_F_MULTI;
+	reply->nlmsg_seq = seq;
+	reply->nlmsg_pid = cps_conf->nl_pid;
+
+	ifam = mnl_nlmsg_put_extra_header(reply, sizeof(*ifam));
+	ifam->ifa_family = family;
+	ifam->ifa_flags = IFA_F_PERMANENT;
+	ifam->ifa_scope = RT_SCOPE_UNIVERSE;
+	ifam->ifa_index = kni_index;
+
+	/*
+	 * The exact meaning of IFA_LOCAL and IFA_ADDRESS depend
+	 * on the address family being used and the device type.
+	 * For broadcast devices (like the interfaces we use),
+	 * for IPv4 we specify both and they are used interchangeably.
+	 * For IPv6, only IFA_ADDRESS needs to be set.
+	 */
+
+	if (family == AF_INET) {
+		mnl_attr_put_u32(reply, IFA_LOCAL, iface->ip4_addr.s_addr);
+		mnl_attr_put_u32(reply, IFA_ADDRESS, iface->ip4_addr.s_addr);
+		ifam->ifa_prefixlen = iface->ip4_addr_plen;
+	} else if (likely(family == AF_INET6)) {
+		mnl_attr_put(reply, IFA_ADDRESS,
+			sizeof(iface->ip6_addr), &iface->ip6_addr);
+		ifam->ifa_prefixlen = iface->ip6_addr_plen;
+	} else {
+		rte_panic("Invalid address family (%hhu) in request while being processed by CPS block in %s\n",
+			family, __func__);
+	}
+}
+
+static int
+rd_getaddr(const struct nlmsghdr *req, const struct cps_config *cps_conf,
+	int *err)
+{
+	char buf[2 * MNL_SOCKET_BUFFER_SIZE];
+	struct mnl_nlmsg_batch *batch;
+	struct net_config *net_conf = cps_conf->net;
+	int family;
+	const char *family_str;
+
+	if (mnl_nlmsg_get_payload_len(req) < sizeof(struct rtgenmsg)) {
+		CPS_LOG(ERR, "Not enough room in CPS GETADDR message from routing daemon in %s\n",
+			__func__);
+		*err = -EINVAL;
+		goto out;
+	}
+
+	family = ((struct rtgenmsg *)mnl_nlmsg_get_payload(req))->rtgen_family;
+
+	/* We don't support AF_UNSPEC to dump both addresses at once. */
+	if (family != AF_INET && family != AF_INET6) {
```

Comment:
Does the kernel interpret `AF_UNSPEC` as a request to dump both addresses at once? If so, we must do so.